java-checks/src/main/java/org/sonar/java/checks/aws/AwsLambdaSyncCallCheck.java

```diff
@@ -0,0 +1,184 @@
+      .names("setInvocationType", "withInvocationType")
+      .addParametersMatcher("java.lang.String").build();
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      getSyncCalls(tree).ifPresent(msgPart -> invokeInvocations.put(tree, msgPart));
+    }
+
+    public Map<MethodInvocationTree, String> getSyncInvokeCalls() {
+      return invokeInvocations;
+    }
+
+    private static Optional<String> getSyncCalls(MethodInvocationTree tree) {
+      if (INVOKE_MATCHERS.matches(tree)) {
+        // INVOKE_MATCHER ensures that there is one argument and it is of type IdentifierTree.
+        IdentifierTree invokeRequest = (IdentifierTree) tree.arguments().get(0);
+
+        // We know there is at least one usage, i.e. the one we just got above.
+        List<IdentifierTree> localUsages = invokeRequest.symbol().usages().stream()
+          .filter(u -> isLocalVariable(u.symbol()) && !u.equals(invokeRequest))
+          .collect(Collectors.toList());
+
+        if (isParameter(invokeRequest.symbol())
+          || localUsages.stream().anyMatch(lu -> isArgumentToACall(lu)
+            || statementSetsAsyncCall(lu))
+          || declarationSetsAsyncCall(invokeRequest)) {
+          return Optional.empty();
+        }
+
+        if (hasLocalVarDeclaration(invokeRequest)) {
+          return Optional.of(MESSAGE);
+        }
+        return Optional.empty();
+      } else {
+        return Optional.empty();
+      }
+    }
+
+    private static boolean isArgumentToACall(IdentifierTree invokeRequest) {
+      return invokeRequest.parent() != null && invokeRequest.parent().is(Tree.Kind.ARGUMENTS);
+    }
+
+    private static boolean hasLocalVarDeclaration(IdentifierTree invokeRequest) {
+      Tree declaration = invokeRequest.symbol().declaration();
+      return (declaration != null && declaration.is(Tree.Kind.VARIABLE) && isLocalVariable(((VariableTree) declaration).symbol()));
+    }
+
+    /**
+     * Returns true if the statement starting at the identifier 'invokeRequest' sets the InvocationType object
+     * to a configuration that sets lambda calls to be async.
+     *
+     * @param invokeRequest the identifier where the statement starts
+     * @return true if statement leads calls to lambdas to be async
+     */
+    private static boolean statementSetsAsyncCall(IdentifierTree invokeRequest) {
+      return callChainSetsAsyncCall(invokeRequest);
+    }
+
+    /**
+     * Returns true if in the call chain starting at 'tree' there is a call that
+     * sets 'InvocationType' to async.
+     *
+     * @param tree an Identifier or a MethodInvocation as starting point of a call chain
+     * @return true if InvocationType is set to async
+     */
+    private static boolean callChainSetsAsyncCall(Tree tree) {
+      Tree treeParent = tree.parent();
+      if (treeParent != null && treeParent.parent() != null &&
+        treeParent.parent().is(Tree.Kind.METHOD_INVOCATION)) {
```

Comment:
Take the following case:
```
      var invokeRequest4 = new InvokeRequest();
      (invokeRequest4).setInvocationType("Event");
      awsLambda.invoke(invokeRequest4);
```
We don't detect the async setter here, as the invocation target is in parentheses, leading to an FP.